gapis/trace/android/trace.go

```diff
@@ -50,12 +71,134 @@ type androidTracer struct {
+			},
+			DurationMs: proto.Uint32(durationMs),
+			DataSources: []*perfetto_pb.TraceConfig_DataSource{
+				{
+					Config: &perfetto_pb.DataSourceConfig{
+						Name: proto.String(gpuRenderStagesDataSourceDescriptorName),
+					},
+				},
+				{
+					Config: &perfetto_pb.DataSourceConfig{
+						Name: proto.String(gpuCountersDataSourceDescriptorName),
+						GpuCounterConfig: &perfetto_pb.GpuCounterConfig{
+							CounterPeriodNs: proto.Uint64(counterPeriodNs),
+							CounterIds:      ids,
+						},
+					},
+				},
+			},
+		},
+	}, nil
 }
 
 func (t *androidTracer) GetDevice() bind.Device {
 	return t.b
 }
 
+func (t *androidTracer) Validate(ctx context.Context) error {
+	ctx = status.Start(ctx, "Android Device Validation")
+	defer status.Finish(ctx)
+	if t.v == nil {
+		return log.Errf(ctx, nil, "No validator found for device %d", t.b.Instance().ID.ID())
+	}
+	d := t.b.(adb.Device)
+	osConfiguration := d.Instance().GetConfiguration()
+	if osConfiguration.GetPerfettoCapability() == nil {
+		return log.Errf(ctx, nil, "No Perfetto Capability found on device %d", d.Instance().ID.ID())
+	}
+	if gpuProfiling := osConfiguration.GetPerfettoCapability().GetGpuProfiling(); gpuProfiling == nil || gpuProfiling.GetGpuCounterDescriptor() == nil {
+		return log.Errf(ctx, nil, "No GPU profiling capabilities found on device %d", d.Instance().ID.ID())
+	}
+
+	// Get ActivityAction
+	packages, _ := d.InstalledPackages(ctx)
+	pkg := packages.FindByName(validationPackageName)
+	if pkg == nil {
+		return log.Errf(ctx, nil, "Package %v not found", validationPackageName)
+	}
+	activityAction := pkg.ActivityActions.FindByName(intentAction, activityName)
+	if activityAction == nil {
+		return log.Errf(ctx, nil, "Activity %v not found in %v", activityName, validationPackageName)
+	}
+
+	// Construct trace config
+	traceOpts, err := traceOptions(ctx, t.v)
+	if err != nil {
+		return log.Err(ctx, err, "Could not get the trace configuration")
+	}
+	process, cleanup, err := perfetto_android.Start(ctx, d, activityAction, traceOpts)
+	if err != nil {
+		cleanup.Invoke(ctx)
+		return log.Err(ctx, err, "Error when start Perfetto tracing.")
+	}
+	defer cleanup.Invoke(ctx)
+
+	// Force to stop the application, ignore any error that happens as it
+	// doesn't affect validation.
+	defer d.ForceStop(ctx, validationPackageName)
+
+	// Start to capture.
+	ctx = status.Start(ctx, "Android Device Validation: Tracing")
+	startSignal, startFunc := task.NewSignal()
+	stopSignal, _ := task.NewSignal()
+	doneSignal, doneFunc := task.NewSignal()
+
+	var buf bytes.Buffer
+	var written int64
+	crash.Go(func() {
+		_, err = process.Capture(ctx, startSignal, stopSignal, &buf, &written)
+		doneFunc(ctx)
+	})
+
+	// TODO(lpy): If we start tracing too early, the data has a lot of noise
+	// at the beginning, need to figure out a sweet spot.
+	startFunc(ctx)
+	if !doneSignal.Wait(ctx) {
+		return log.Err(ctx, err, "Fail to wait for done signal from Perfetto.")
```

Comment:
would need a `status.Finish` here, also the `status.Start` below would need the context from before the above `status.Start`.

You can use
```
status.Do(ctx, "thing you're doing", func(ctx context.Context) {
   // put your code here
})
```
It will call the passed function in between start and finish.

I can explain a bit about the quirks of go's `context.Context` on Monday.